WIP: Export IFW (Beta)

podsvirov opened the pull request:
Add export to binary crossplatform repository/installer
with GUI based on QtIFW:
http://doc.qt.io/qtinstallerframework/ifw-overview.html

For correct operation of these changes,
you must use the corrected QtIFW:
https://codereview.qt-project.org/#/c/203958

UnaNancyOwen committed:
Add Kinect driver to OpenNI2 port

Add Kinect driver to OpenNI2 port.
OpenNI2 port generate Kinect driver when Kinect SDK has been installed
in your system.

UnaNancyOwen committed:
Add Kinect SDK v2.x port

Add Kinect for Windows SDK v2.x port.

UnaNancyOwen committed:
Dynamic path convert

Dynamic path convert of Kinect SDK directory in patch file.

podsvirov committed:
WIP: Export IFW

Add export to binary crossplatform repository/installer
with GUI based on QtIFW:
http://doc.qt.io/qtinstallerframework/ifw-overview.html

For correct operation of these changes,
you must use the corrected QtIFW:
https://codereview.qt-project.org/#/c/203958

msftclas commented:
@__podsvirov__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla.microsoft.com.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
Microsoft Pull Request Bot_

msftclas commented:
__@podsvirov__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, Microsoft Pull Request Bot_

podsvirov commented:
Hello everyone interested!
Today I have something to show you.
This is a stand-alone stand-alone installer created based on this change request.
Installer:
[vcpkg-export-ifw-standalone.exe](http://ifw.podsvirov.pro/vcpkg/demo/2017-08-30/vcpkg-export-ifw-standalone.exe)
Screenshot (Please, do not pay attention to partial localization): ![vcpkg-export-ifw.png](http://ifw.podsvirov.pro/vcpkg/demo/2017-08-30/vcpkg-export-ifw.png)
I will be glad to receive feedback and constructive suggestions.

podsvirov commented:
Hello @alexkaratarakis and @ras0219-msft,
what your think about add this feature?
What is the probability of accepting my changes after revision and refactoring?

drdanz commented on toolsrc/src/commands_export.cpp:
```diff
@@ -348,10 +418,227 @@ namespace vcpkg::Commands::Export
+            line += "<Package>";
+            lines.push_back(line); line = skip;
+            line += "<DisplayName>";
+            line += "Intergation";
```
There is a typo here... Intergation -> Integration (same typo also a few lines later in the comment)

podsvirov replied:
Good catch. Thank you!

drdanz reviewed (commented):
It looks like a great addition to me...
I did not review this code, I just noticed this typo from the screenshots.

podsvirov commented:
QtIFW [CR203958/3](https://codereview.qt-project.org/#/c/203958/3) has updated.
Previousle option `AllowDashInName` extended and renamed to `AllowAllInNameAndVersion`.
Updated standalone installer:
[vcpkg-export-ifw-standalone.exe](http://ifw.podsvirov.pro/vcpkg/demo/2017-08-31/vcpkg-export-ifw-standalone.exe)
And I'm glad to announce the online installer:
[vcpkg-export-ifw-online.exe](http://ifw.podsvirov.pro/vcpkg/demo/2017-08-31/vcpkg-export-ifw-online.exe)
New screenshot:
 ![vcpkg-export-ifw.png](http://ifw.podsvirov.pro/vcpkg/demo/2017-08-31/vcpkg-export-ifw-debian-kde.png)
Anyone can test these installers.
I will be glad to receive feedback and constructive suggestions.

podsvirov commented:
@alexkaratarakis, ping :-)
In these changes, I need to generate the files in `XML` format. Please tell me how to do this most effectively within the `vcpkg` project?

glachancecmaisonneuve committed:
find_acquire_program fix for python2
python2 version bump

ras0219-msft commented:
Sorry for the delay! This is really cool and we would love to merge it in with some more polish!

For the time being, we want to avoid external library dependencies to avoid the horrible bootstrapping problem. If there aren't too many fields, it may be clean enough to synthesize with `Strings::format`[1]. Alternatively, we sometimes just use regex as a glorified replace-all [2].

Given the significant size of the XML generation code, I think it would be best to break it into a separate function or even another file.

[1] https://github.com/Microsoft/vcpkg/blob/c2a735f6bee792d2419b0616eaa84abef91536e6/toolsrc/src/metrics.cpp#L171-L206
[2] https://github.com/Microsoft/vcpkg/blob/c2a735f6bee792d2419b0616eaa84abef91536e6/toolsrc/src/metrics.cpp#L171-L206

podsvirov commented:
Hello @ras0219-msft, thanks for approving my ideas. I hope that these changes will be useful for many developers.
What about the time, then we all are busy, especially if it's about open source development (as a hobby, for me now).
Recently, I continue to work on the necessary changes in the QtIFW and [CR203958/9](https://codereview.qt-project.org/#/c/203958/9) has updated. Previousle option  `AllowAllInNameAndVersion` now do not need at all.
@ras0219-msft, I will listen to your recommendations and try to make my code better.

podsvirov commented:
@ras0219-msft, today I have move my code to another files. If somewere is wrong, please correct me.

ras0219-msft commented on toolsrc/src/commands_export.cpp:
```diff
@@ -305,7 +315,11 @@ namespace vcpkg::Commands::Export
+        std::string export_id = create_export_id();
+        if (ifw)
+        {
+            export_id = "vcpkg-export"; // TODO: Remove after debugging
```
TODO: Remove after debugging

ras0219-msft commented on toolsrc/src/commands_export.cpp:
```diff
@@ -327,10 +343,21 @@ namespace vcpkg::Commands::Export
+            if (ifw)
+            {
+                // Export real package and return data dir for installation
+                spec_exported_dir_path = IFW::export_real_package(raw_exported_dir_path, action, fs);
```
How does this handle multiple export formats at the same time (i.e. `--ifw --nuget --raw`)? I think either the `from_destination_root()` call needs to be duplicated _or_ you could add a check that requires `--ifw` to not be passed with any other export types.

ras0219-msft commented on toolsrc/src/commands_export.cpp:
```diff
@@ -351,13 +378,29 @@ namespace vcpkg::Commands::Export
 
+        if (ifw)
+        {
+            // Unigue packages
```
Typo

ras0219-msft commented on toolsrc/src/commands_export.cpp:
```diff
@@ -351,13 +378,29 @@ namespace vcpkg::Commands::Export
+            // Integration
+            IFW::export_integration(raw_exported_dir_path, fs);
+            // Configuration
+            IFW::export_config(raw_exported_dir_path, maybe_ifw_repository_url.value_or(""), fs);
```
It looks like the IFW path really isn't able to reuse much or anything from the main loop -- it might make sense to completely extract it from the loop and just have
```cpp
IFW::export(export_plan, raw_exported_dir_path, fs);
```

Then, we'd completely skip the main loop when only running `--ifw`.

Note that I do think the functions should probably stay separated, but they'd be private to the IFW implementation resulting in a "cleaner" public interface.

ras0219-msft commented on toolsrc/src/commands_export.cpp:
```diff
@@ -417,7 +460,7 @@ With a project open, go to Tools->NuGet Package Manager->Package Manager Console
         }
 
-        if (!raw)
+        if (!raw && !ifw)
```
I assume this was for debugging?

ras0219-msft committed:
[vcpkg-export-ifw] Use template approach for xml instead of line-by-line

ras0219-msft commented:
I pushed a commit that switches to using `sprintf`-style templates instead of line-by-line, which makes the code _significantly_ more compact and (at least for me) easier to read.

podsvirov commented:
Hello @ras0219-msft, thanks for your comments and commit. Now I will correct my mistakes.

podsvirov committed:
[vcpkg-export-ifw] Separate IFW loop

Separate IFW loop compatible with main export loop
Fixed mistakes in templates
Set current date to ReleaseDate tag

podsvirov commented:
@ras0219-msft hello again, I has improve code (add 1 commit after you) and rebase result to current `master`.
Please review again.

alexkaratarakis committed:
[zlib] Patch to not build more things than needed

JonnyPtn committed:
Initial enet port

JonnyPtn committed:
Initial GTS port

podsvirov committed:
[vcpkg-export-ifw] Usage QtIFW tools

Download and use tools to make repository and installer

podsvirov commented:
Hello @ras0219-msft and all!
Now you can try create installer:
```
.\vcpkg.exe export libpng:x64-windows --ifw
```
Or to create repository and online installer
```
.\vcpkg.exe export libpng:x64-windows --ifw --ifw-repository-url=http://your.domain/ifw/repo
```
Please, try this and report the results.

podsvirov commented:
But I have problem with first run (when download and unzip tools):
```

J:\build>.\vcpkg-export-ifw.bat

J:\build>.\vcpkg\vcpkg.exe export libpng:x86-windows libpng:x64-windows c-ares:x
86-windows c-ares:x64-windows --ifw --ifw-packages-directory-path=..\vcpkg-expor
t-ifw-packages --ifw-repository-directory-path=..\vcpkg-export-ifw-repository  -
-ifw-repository-url=file:///J:/build/vcpkg-export-ifw-repository --ifw-configura
tion-file-path=..\vcpkg-export-ifw-configuration.xml --ifw-installer-file-path=.
.\vcpkg-export-ifw-installer.exe
The following packages are already built and will be exported:
    c-ares:x86-windows
    c-ares:x64-windows
    libpng:x86-windows
    libpng:x64-windows
  * zlib:x86-windows
  * zlib:x64-windows
Additional packages (*) need to be exported to complete this operation.
Exporting packages ../vcpkg-export-ifw-packages...
Exporting package zlib:x86-windows...
Exporting package zlib:x86-windows... done
Exporting package libpng:x86-windows...
Exporting package libpng:x86-windows... done
Exporting package zlib:x64-windows...
Exporting package zlib:x64-windows... done
Exporting package libpng:x64-windows...
Exporting package libpng:x64-windows... done
Exporting package c-ares:x86-windows...
Exporting package c-ares:x86-windows... done
Exporting package c-ares:x64-windows...
Exporting package c-ares:x64-windows... done
Exporting packages ../vcpkg-export-ifw-packages... done
Generating configuration ../vcpkg-export-ifw-configuration.xml...
Generating configuration ../vcpkg-export-ifw-configuration.xml... done.
Expected dependency downloaded path to be J:\build\vcpkg\downloads\QtInstallerFr
amework-win-x86\bin\installerbase.exe, but was ?У?З?ТA╗?ТNЧJ:\build\vcpkg\downlo
ads\QtInstallerFramework-win-x86\bin\installerbase.exe

J:\build>.\vcpkg-export-ifw.bat

J:\build>.\vcpkg\vcpkg.exe export libpng:x86-windows libpng:x64-windows c-ares:x
86-windows c-ares:x64-windows --ifw --ifw-packages-directory-path=..\vcpkg-expor
t-ifw-packages --ifw-repository-directory-path=..\vcpkg-export-ifw-repository  -
-ifw-repository-url=file:///J:/build/vcpkg-export-ifw-repository --ifw-configura
tion-file-path=..\vcpkg-export-ifw-configuration.xml --ifw-installer-file-path=.
.\vcpkg-export-ifw-installer.exe
The following packages are already built and will be exported:
    c-ares:x86-windows
    c-ares:x64-windows
    libpng:x86-windows
    libpng:x64-windows
  * zlib:x86-windows
  * zlib:x64-windows
Additional packages (*) need to be exported to complete this operation.
Exporting packages ../vcpkg-export-ifw-packages...
Exporting package zlib:x86-windows...
Exporting package zlib:x86-windows... done
Exporting package libpng:x86-windows...
Exporting package libpng:x86-windows... done
Exporting package zlib:x64-windows...
Exporting package zlib:x64-windows... done
Exporting package libpng:x64-windows...
Exporting package libpng:x64-windows... done
Exporting package c-ares:x86-windows...
Exporting package c-ares:x86-windows... done
Exporting package c-ares:x64-windows...
Exporting package c-ares:x64-windows... done
Exporting packages ../vcpkg-export-ifw-packages... done
Generating configuration ../vcpkg-export-ifw-configuration.xml...
Generating configuration ../vcpkg-export-ifw-configuration.xml... done.
Generating repository ../vcpkg-export-ifw-repository...
Generating repository ../vcpkg-export-ifw-repository... done.
Generating installer ../vcpkg-export-ifw-installer.exe...
Generating installer ../vcpkg-export-ifw-installer.exe... done.

To use the exported libraries in CMake projects use:
    "-DCMAKE_TOOLCHAIN_FILE=@RootDir@/src/vcpkg/scripts/buildsystems/vcpkg.cmake
"


J:\build>
```
But second run passed.
The problem with some unnecessary symbols... Please help me.

jasjuang committed:
new port libfreenect2

podsvirov committed:
[vcpkg-export-ifw] Some improvements

Improvements:
- fix typos;
- remove outdated repository directory.

podsvirov commented:
And I'm glad to announce the beta online installer:
[vcpkg-export-ifw-online.exe](http://ifw.podsvirov.pro/vcpkg/beta/vcpkg-export-ifw-installer.exe)

Current packages:
- grpc:x86-windows  and grpc:x64-windows
- glfw3:x86-windows and glfw3:x64-windows
- libpng:x86-windows and libpng:x64-windows
- c-ares:x86-windows and c-ares:x64-windows
- plus depend packages...

PS: Is there anyone here?

alexkaratarakis committed:
Extract function for duplicated code

alexkaratarakis committed:
Place line closer to usage

TheCycoONE committed:
Dynamically link lua.dll to lua.exe

For lua.exe to load C modules which are linked against lua, it must itself be
linked against lua, otherwise loading the module fails.

alexkaratarakis committed:
Fix: Don't break before finding all VS2017 instances

alexkaratarakis committed:
[Toolset selection] Use VS2017 vcvarsall for v140 if available

jasjuang committed:
Merge remote-tracking branch 'upstream/master'

ras0219-msft committed:
[abseil] Initial commit. Highly experimental.

ras0219-msft committed:
[abseil] Improve warning messages.

alexkaratarakis committed:
Merge pull request #1876 from JonnyPtn/enet-port

Initial enet port

alexkaratarakis committed:
Merge pull request #1877 from JonnyPtn/gts-port

Initial GTS port

alexkaratarakis committed:
Merge pull request #1882 from jasjuang/master

new port libfreenect2

alexkaratarakis committed:
Merge pull request #1341 from UnaNancyOwen/fix_openni2

Add Kinect driver to OpenNI2 port

alexkaratarakis committed:
Revert "Extract function for duplicated code"

This reverts commit eca068e4c5673b888d6470b42d6b5b44b21c4007.

alexkaratarakis committed:
Merge pull request #1884 from TheCycoONE/lua_lib

Dynamically link lua.dll to lua.exe

alexkaratarakis committed:
[opengl] Rename portfile variable

jasjuang committed:
Merge remote-tracking branch 'upstream/master'

jasjuang committed:
new port: exiv2

JesseTG committed:
Add a port for spirit-po

- It's header-only, so this makes installation easy

JesseTG committed:
Crap

amincheloh committed:
[vcpkg-docs] Update CONTROL file

Build-Depends change qualifier from brackets to parentheses. see #1582

alexkaratarakis committed:
size_t instead of int

alexkaratarakis committed:
Introduce Strings::case_insensitive_ascii_starts_with();

Mixaill committed:
[libexif] add version 0.6.21

jasjuang committed:
add in expat dependency

jasjuang committed:
update commit to master for cmake config support

alexkaratarakis committed:
Merge pull request #1897 from Mixaill/libexif-0621

[libexif] add version 0.6.21

alexkaratarakis committed:
Merge pull request #1894 from amincheloh/patch-1

[vcpkg-docs] Update CONTROL file

alexkaratarakis committed:
Merge pull request #1893 from JesseTG/jtg/spirit-po

Add a port for spirit-po

alexkaratarakis committed:
Merge pull request #1892 from jasjuang/exiv2

new port: Exiv2

ShinNoNoir committed:
[ffmpeg] Deal with case-sensitive env vars

When behind a corporate proxy, one often needs to specify `HTTP_PROXY` and `HTTPS_PROXY` for some command line tools to work properly. However, `pacman` seems to rely on the lowercase equivalent environment variables. In a Windows command prompt environment, it is not possible to set both since Windows environment variables are not case-sensitive. As a workaround, this build script checks for the existence of HTTP_PROXY and HTTPS_PROXY. If they exist, they are exported as lowercase variables.

Ghabry committed:
Add xmplite 4.4.1

Ghabry committed:
Add Wildmidi 0.4.1

UnaNancyOwen committed:
Add Dark support

Add Dark support to vcpkg_find_acquire_program().

ras0219-msft committed:
Merge pull request #1829 from glachancecmaisonneuve/find_acquire_python2

find_acquire_program (python2)

ras0219-msft committed:
Merge pull request #1904 from UnaNancyOwen/add_wix

Add Dark support

Commit:
[FreeXL] initial port

Commit:
[ReadOSM] initial port

ras0219-msft committed:
[libxmp-lite] Add INSTALL_HEADERS cmake macro

ras0219-msft committed:
Merge pull request #1901 from Ghabry/add-xmplite

Add xmplite 4.4.1

ras0219-msft committed:
[vcpkg] Fix use of UNICODE-sensitive Windows APIs

ras0219-msft committed:
Merge pull request #1900 from ShinNoNoir/patch-2

[ffmpeg] Deal with case-sensitive env vars

alexkaratarakis committed:
Merge pull request #1907 from alex85k/readosm

[ReadOSM] initial port

alexkaratarakis committed:
Add message when downloading cmake/git/nuget

alexkaratarakis committed:
Show warning instead of failing if port cannot be parsed

ras0219-msft committed:
[proj][proj4] Rename proj to proj4. Leave empty forwarder behind.

alexkaratarakis committed:
Extract local variable

alexkaratarakis committed:
Merge pull request #1906 from alex85k/freexl

[FreeXL] initial port

alexkaratarakis committed:
Merge pull request #1902 from Ghabry/add-wildmidi

Add Wildmidi 0.4.1

alexkaratarakis committed:
Replace deprecated std::codecvt/std::wstring_convert. Fixes building with v141

alexkaratarakis committed:
Change size_t to int (which is the actual return type)

alexkaratarakis committed:
Suppress 4768 warning from shlobj.h

MatthewPowley committed:
[telnetpp] Initial port for 'Telnet++', a C++ 14 implementation of the Telnet Session Layer protocol

* Initial commit for a port of Telnet++ (telnetpp)
* Upstream source: https://github.com/KazDragon/telnetpp
* Depends on Boost and Zlib (Additional dependency on GTest due to upstream authoring error)

UnaNancyOwen committed:
Extract WiX installer using Dark

Extract WiX installer using Dark.
It will be standalone extract files from installer of Kinect SDK 2.x
even if Kinect SDK 2.x is not installed in user system.

Commit:
[ReadOsm][FreeXL] fix spacebar-in-path problems

Commit:
[geos] allow geos_c static library cretation (libgeos_c.lib) v2

Commit:
[libspatialite] initial port

alexkaratarakis committed:
Merge pull request #1915 from alex85k/fix-spacebar

[ReadOsm][FreeXL] fix spacebar-in-path problems

alexkaratarakis committed:
Merge pull request #1917 from alex85k/libspatialite

[libspatialite] initial port

alexkaratarakis committed:
Merge pull request #1913 from MatthewPowley/Telnetpp

[telnetpp] Initial port for 'Telnet++'

ras0219-msft committed:
Merge pull request #1905 from alex85k/master

[geos] allow geos_c static library cretation (libgeos_c.lib)

ras0219-msft committed:
[geos] Bump version for previous PR

ras0219-msft committed:
[vcpkg] Add more operator== to CStringView. Uppercase Span to follow naming convention.

ras0219-msft committed:
[gtest] Silence TR1 deprecation warning

podsvirov commented:
CppCon is done... Ping :-)

atkawa7 committed:
[geos] remove build_cmake

atkawa7 committed:
[geos] bump version

UnaNancyOwen committed:
Extract WiX installer using Dark

Extract WiX installer using Dark.
It will be standalone extract files from installer of Kinect SDK 1.x
even if Kinect SDK 1.x is not installed in user system.

UnaNancyOwen committed:
Change to refer Kinect SDK 1.x that installed using vcpkg port

Fix according to changes of Kinect SDK 1.x port.
Change to refer Kinect SDK 1.x that installed using vcpkg port.
It will be always generate the Kinect SDK 1.x driver, even when Kinect
SDK 1.x is not installed in user system.

ras0219-msft committed:
Merge pull request #1923 from atkawa7/patch-1

[geos] remove build_cmake

alexkaratarakis committed:
Merge pull request #1414 from UnaNancyOwen/add_kinectsdk2

Add Kinect SDK v2.x port

alexkaratarakis committed:
Merge pull request #1925 from UnaNancyOwen/fix_kinectsdk1

Extract Kinect SDK 1.x Installer using Dark

alexkaratarakis committed:
Merge pull request #1926 from UnaNancyOwen/fix_openni2

Change to refer Kinect SDK 1.x that installed using vcpkg port

ras0219-msft committed:
Merge branch 'export-ifw' of https://github.com/podsvirov/vcpkg into podsvirov-export-ifw

ras0219-msft committed:
[vcpkg] Reformat and fix headers.

ras0219-msft committed:
[vcpkg] Reformat and refactor to reduce function size

ras0219-msft committed:
[vcpkg] Refactor out implication in option parsing for export

ras0219-msft commented:
Thanks again for all this work, it's really incredible!

I've merged for now (with a few refactorings/housekeepings of my own); I know you have some follow-up improvements and I'd be happy to handle them through separate PRs!

podsvirov commented:
@ras0219-msft, thank you for merging my changes. It was really unexpected for me now, but I'm very glad that it happened. Of course I have plans for further improvement and I will do this work consistently on.

podsvirov commented:
@ras0219-msft, the problem of the first export is still relevant to me:
```
A suitable version of installerbase was not found (required v3.1.81). Downloadin
g portable installerbase v3.1.81...
Expected dependency downloaded path to be J:\build\vcpkg\downloads\QtInstallerFr
amework-win-x86\bin\installerbase.exe, but was ?У?З?ТA¬?ТNЧJ:\build\vcpkg\downlo
ads\QtInstallerFramework-win-x86\bin\installerbase.exe
```
Unexpected simbols ``?У?З?ТA¬?ТNЧJ``...
But the tools are downloaded and unpacked normally and on subsequent attempts work as expected.

ras0219-msft commented:
Hmm, we had some issues with byte-order marks before and were able to solve them with [1].

Could you put a breakpoint there, repro the issue, and let us know what the exact byte values are at the beginning of the response?

[1] https://github.com/Microsoft/vcpkg/blob/d8507cd159f8de6ae174cae9894c926f75fe50d7/toolsrc/src/vcpkg_System.cpp#L180-L188

podsvirov commented:
I reproduced this and here is my result (I see 5 extra characters):
![vcpkg-export-ifw-unexpected-symbols](https://user-images.githubusercontent.com/8400779/31256827-7af439d6-aa3d-11e7-978f-58a107f92e83.png)

ras0219-msft commented:
Thanks, that's really helpful! It appears you're getting _two_ byte order marks (because your bytes weren't sufficiently ordered, I guess? 😋).

I've pushed 9b0c2cb2 to master which changes that function to loop until we no longer find any BOMs; I think this will solve your issue.

podsvirov commented:
@ras0219-msft thanks, it seems that now everything works as it should.
